crates/sui-core/src/consensus_adapter.rs

```diff
@@ -284,6 +354,31 @@ impl ConsensusAdapter {
         processed_waiter
             .await
             .expect("Storage error when waiting for consensus message processed");
+        let send_end_of_publish =
+            if let ConsensusTransactionKind::UserTransaction(certificate) = &transaction.kind {
+                let reconfig_guard = self.authority.get_reconfig_state_read_lock_guard();
+                // note that pending_certificates lock is always acquired *after* reconfiguration lock
+                // acquiring locks in different order might lead to deadlocks
+                let mut pending_certificates = self.pending_certificates.lock();
+                pending_certificates.remove(certificate.digest().as_ref());
+                if reconfig_guard.is_reject_user_certs() {
```

Comment:
Why not just call `!reconfig_guard.should_accept_user_certs`?
I find the name `is_reject_user_certs` a bit confusing

Reply:
`should_accept_user_certs` is not equivalent call though. In particular `!reconfig_guard.should_accept_user_certs` as you suggested would return true in `RejectAllCerts` state, but we don't need to send EndOfPublish in this state

Why is the name confusing? This seem pretty standard when enum has element `RejectUserCerts`, there is bool function `is_<enum_element_name`. See `Option::is_none`, `Option::is_some`, `Error::is_ok`, etc.

Reply:
If it's in RejectAllCerts state, we won't have reached this line of code? (because it would have been rejected early on)

Reply:
This is probably true, but why make some implicit assumption(the state can not be reached so we ignore it), when we can make explicit check?

Also, this would not remove `is_reject_user_certs` function, there is another usage for it

Reply:
OK that's fair. But let's:
- Add a debug_assert here that it should be in the reject all state
- Add some comments

Reply:
I actually think we can't add assert here - since this is executed asynchronously to consensus task you might in theory get signal here while state already transitioned to RejectAllCerts. This race condition does not need special handling, but we can't assert it

Reply:
Added comment